runbotics/runbotics-scheduler/src/scheduler-database/process-credential/process-credential.service.ts

```diff
@@ -0,0 +1,129 @@
+
+@Injectable()
+export class ProcessCredentialService {
+    constructor(
+        @InjectRepository(ProcessCredential)
+        private readonly processCredentialRepository: Repository<ProcessCredential>,
+        @InjectRepository(Credential)
+        private readonly credentialRepository: Repository<Credential>,
+        private readonly processService: ProcessService,
+    ) {}
+
+    async findAllByProcessId(processId: number, user: UserEntity) {
+        await this.processService.checkAccessAndGetById(processId, user);
+
+        return this.processCredentialRepository.find({
+            where: {
+                process: { id: processId },
+                credential: {
+                    tenantId: user.tenantId,
+                }
+            },
+            relations: ['credential.template', 'credential.createdBy', 'credential.collection']
+        }).then(this.formatProcessCredentials);
+    }
+
+    async create(processCredentialDto: CreateProcessCredentialDto, user: UserEntity) {
+        const process = await this.processService.checkAccessAndGetById(
+            +processCredentialDto.processId, user
+        );
+
+        const credential = await this.credentialRepository.findOneByOrFail({
+            id: processCredentialDto.credentialId, tenantId: user.tenantId
+        }).catch(() => {
+            throw new NotFoundException('Cannot find credential with provided id');
+        });
+
+        const lastCredentialOrder: number = await this.processCredentialRepository
+            .createQueryBuilder('pc')
+            .select('MAX(pc.order)', 'max')
+            .innerJoin('pc.credential', 'credential')
+            .innerJoin(
+                'credential.template', 'template',
+                'template.name = :templateName', { templateName: processCredentialDto.templateName }
+            ).where('pc.processId = :processId', { processId: processCredentialDto.processId })
+            .groupBy('template.name')
+            .getRawOne()
+            .then(res => res.max)
+            .catch(() => 0);
+
+        const newProcessCredential = new ProcessCredential();
+        newProcessCredential.process = process;
+        newProcessCredential.credential = credential;
+        newProcessCredential.order = lastCredentialOrder + 1;
+        await this.processCredentialRepository.insert(newProcessCredential);
+    }
+
+    async delete(id: string, user: UserEntity) {
+        const processCredential = await this.processCredentialRepository.findOneOrFail({
+            where: { id, credential: { tenantId: user.tenantId } },
+            relations: ['process', 'credential.template']
+        }).catch(() => {
+            throw new NotFoundException();
+        });
+
+        const processId = processCredential.process.id;
+
+        await this.processService.checkAccessAndGetById(
+            processId, user
+        );
+
+        await this.processCredentialRepository.delete(id);
+
```

Comment:
1. this should be done in a transaction
2. 
```
const credentialsToSave = await repo.getProcess(id);
credentialsToSave.sort(a,b => a.order - b.order).filter(c => c.order < x).map((c, idx) => {...c, order: idx))
repo.update(credentialsToSave);
```

simpler code, worth the small overhead of fetching imo

Reply:
I had a feeling that transaction would fit here. I will think about those.